iati/core/rulesets.py

```diff
@@ -578,19 +739,26 @@ def is_valid_for(self, dataset):
 
         Raises:
             AttributeError: When an argument is given that does not have the required attributes.
-            XPathEvalError(lxml.etree.XPathEvalError): When no valid XPath is available.
 
         Todo:
             Consider better methods for specifying which elements in the tree contain non-permitted duplication, such as bucket sort.
 
         """
+        context_elements = self._find_context_elements(dataset)
         original = list()
         unique = set()
 
-        for path in self.paths:
-            results = dataset.xml_tree.xpath(path)
-            for result in results:
-                original.append(result.text)
-                unique.add(result.text)
+        for context_element in context_elements:
+            if self._evaluate_condition(context_element):
+                return None
+            for path in set(self.paths):
+                results = context_element.xpath(path)
+                strings_to_check = self._extract_text_from_element_or_attribute(results)
+                for string_to_check in strings_to_check:
+                    original.append(string_to_check)
+                    unique.add(string_to_check)
+
+        if len(original) != len(unique):
```

Comment:
This check should be performed within each context, then the set/list reset.

Reply:
(remember to create a failing test before changing this!)

Reply:
done